autonomous/shoot_move_shoot.py

```diff
@@ -0,0 +1,235 @@
+import math
+
+from wpilib import controller
+from wpilib import geometry
+from wpilib import trajectory
+from wpilib.trajectory import constraint
+from magicbot import AutonomousStateMachine, state
+
+from components.chassis import Chassis
+from components.indexer import Indexer
+from components.shooter import Shooter
+from controllers.shooter import ShooterController
+
+
+def to_pose(x: float, y: float, heading: float) -> geometry.Pose2d:
+    """
+    Convert inputs into a wpilib pose object
+    """
+    rotation = geometry.Rotation2d(heading)
+    return geometry.Pose2d(x, y, rotation)
+
+
+class ShootMoveShootBase(AutonomousStateMachine):
+
+    shooter_controller: ShooterController
+
+    chassis: Chassis
+    indexer: Indexer
+    shooter: Shooter
+
+    # has_zeroed: bool
+
+    TARGET_POSITION = geometry.Translation2d(0, 0)
+
+    def __init__(self) -> None:
+        super().__init__()
+        self.controller = controller.RamseteController()
+        tolerance = to_pose(0.1, 0.1, math.pi / 18)
+        self.controller.setTolerance(tolerance)
+        self.trajectory_config = trajectory.TrajectoryConfig(
+            maxVelocity=1, maxAcceleration=1
+        )
+        self.gen = trajectory.TrajectoryGenerator()
+        self.trajectory_num = 0
+
+    def setup(self):
+        self.trajectory_config.setKinematics(self.chassis.kinematics)
+        self.trajectory_config.addConstraint(
+            constraint.DifferentialDriveVoltageConstraint(
+                self.chassis.ff_calculator, self.chassis.kinematics, 10
+            )
+        )
+        self.end_ranges = []
+        self.paths = []
+        for i in range(self.trajectory_max):
```

Comment:
We should probably refactor this at some point so it's impossible to get all these out of sync.